src/lazy/binary/lazy_reader.rs

```diff
@@ -0,0 +1,222 @@
+///
+/// // Visit the values in the list
+/// let mut sum = 0;
+/// for lazy_value in &lazy_list {
+///     // Read each lazy value in the lazy list as an int (i64) and
+///     // add it to the running total
+///     sum += lazy_value?.read()?.expect_i64()?;
+/// }
+///
+/// assert_eq!(sum, 60);
+///
+/// // Note that we can re-read any of the lazy values. Here we'll step into the list again and
+/// // read the first child value.
+/// let first_int = lazy_list.iter().next().unwrap()?.read()?.expect_i64()?;
+/// assert_eq!(first_int, 10);
+///
+///# Ok(())
+///# }
+/// ```
+pub struct LazyReader<'data> {
+    system_reader: LazySystemReader<'data>,
+}
+
+impl<'data> LazyReader<'data> {
+    pub fn new(ion_data: &'data [u8]) -> IonResult<LazyReader<'data>> {
+        if ion_data.len() < IVM.len() {
+            return decoding_error("input is too short to be recognized as Ion");
+        } else if ion_data[..IVM.len()] != IVM {
+            return decoding_error("input does not begin with an Ion version marker");
+        }
+
+        let system_reader = LazySystemReader::new(ion_data);
+        Ok(LazyReader { system_reader })
+    }
+
+    /// Returns the next top-level value in the input stream as `Ok(Some(lazy_value))`.
+    /// If there are no more top-level values in the stream, returns `Ok(None)`.
+    /// If the next value is incomplete (that is: only part of it is in the input buffer) or if the
+    /// input buffer contains invalid data, returns `Err(ion_error)`.
+    pub fn next<'top>(&'top mut self) -> IonResult<Option<LazyValue<'top, 'data>>> {
+        self.system_reader.next_value()
+    }
+
+    /// Like [`Self::next`], but returns an `IonError` if there are no more values in the stream.
+    pub fn expect_next<'top>(&'top mut self) -> IonResult<LazyValue<'top, 'data>> {
+        self.next()?
+            .ok_or_else(|| decoding_error_raw("expected another top-level value"))
+    }
+}
+
+pub struct LazyElementIterator<'iter, 'data> {
+    lazy_reader: &'iter mut LazyReader<'data>,
+}
+
+impl<'iter, 'data> Iterator for LazyElementIterator<'iter, 'data> {
+    type Item = IonResult<Element>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.lazy_reader.next() {
+            Ok(None) => None,
+            Ok(Some(lazy_value)) => Some(lazy_value.try_into()),
+            Err(e) => Some(Err(e)),
+        }
+    }
+}
+
+impl<'data> ElementReader for LazyReader<'data> {
+    type ElementIterator<'a> = LazyElementIterator<'a, 'data> where Self: 'a,;
+
+    fn read_next_element(&mut self) -> IonResult<Option<Element>> {
+        let lazy_value = match self.next()? {
+            None => return Ok(None),
+            Some(lazy_value) => lazy_value,
+        };
+        let element: Element = lazy_value.try_into()?;
+        Ok(Some(element))
+    }
+
+    fn elements(&mut self) -> Self::ElementIterator<'_> {
+        LazyElementIterator { lazy_reader: self }
+    }
+}
```

Comment:
Oh yeah, I see why you had trouble integrating with the `Element` round trip tests.